src/prefect/_internal/pydantic/v2_validated_func.py

```diff
@@ -24,13 +26,16 @@
 class V2ValidatedFunction(ValidatedFunction):
     def create_model(
         self,
-        fields: Dict[str, Any],
+        fields: dict[str, Any],
         takes_args: bool,
         takes_kwargs: bool,
-        config: ConfigDict,
+        config: "ConfigType",
     ) -> None:
         pos_args = len(self.arg_mapping)
 
+        if TYPE_CHECKING:
+            assert isinstance(config, Mapping)
+
```

Comment:
This is confusing me. Why can't the `config` parameter be annotated with `Mapping[something, something]` here?

Reply:
I did this because the baseclass `ValidatedFunction` has this signature

```python
ConfigType = Union[None, Type[Any], Dict[str, Any]]

class ValidatedFunction:
    def __init__(self, function: 'AnyCallableT', config: 'ConfigType')
```

so when I tried that it was upset with me that I was incongruently overriding the annotation

open to suggestions if you think there's a better approach!

Reply:
Right, that's the context I was missing. So this is code adapting a v1 decorator to work with Pydantic v2.

I had a quick look at [how the original code uses this](https://github.com/pydantic/pydantic/blob/main/pydantic/v1/decorator.py#L214-L218):

```python
        if not TYPE_CHECKING:  # pragma: no branch
            if isinstance(config, dict):
                CustomConfig = type('Config', (), config)  # noqa: F811
            elif config is not None:
                CustomConfig = config  # noqa: F811
```

So the original annotation allows for `None` to mean "empty", a dictionary to form a namespace of a constructed Config class, or a Config class on its own.

You can't indeed just change the method signature because the rest of the class implementation is unchanged and it would rely on this method not to have changed its interface, more or less.

So, I'd accept None (treated as an empty dict) and put in an _unguarded_ check, not behind a `TYPE_CHECKING` check, that the value is a dictionary. And I'd raise a `TypeError` if it's anything else.

You need this because the type annotations for constructing this class are and always will be misleading.

Alternatively, move this class to a private name and put a utility function in its place that has the right signature, and have that utility function construct this class. Only then would a `TYPE_CHECKING`-only assertion be enough.